Integration test for disable default cni flag (#113)
canonical/cluster-api-bootstrap-provider-microk8s: integration/README.md

The integration/e2e tests have the following prerequisites:

  * an environment variable `CLUSTER_MANIFEST_FILE` pointing to the cluster manifest. Cluster manifests can be produced with the help of the templates found under [`templates`](../templates). For example:
  ```bash 
export AWS_REGION=us-east-1
export AWS_SSH_KEY_NAME=capi
export CONTROL_PLANE_MACHINE_COUNT=3
export WORKER_MACHINE_COUNT=3
export AWS_CREATE_BASTION=false
export AWS_PUBLIC_IP=false
export AWS_CONTROL_PLANE_MACHINE_FLAVOR=t3.large
export AWS_NODE_MACHINE_FLAVOR=t3.large
export CLUSTER_NAME=test-ci-cluster
clusterctl generate cluster ${CLUSTER_NAME} --from "templates/cluster-template-aws.yaml" --kubernetes-version 1.27.0 > cluster.yaml
export CLUSTER_MANIFEST_FILE=$PWD/cluster.yaml
  ```

> > NOTE: AWS_SSH_KEY_NAME is name of ssh key in AWS that you are plan to use, if you don't have one yet refer

Reviewer: nit: the name

Reviewer 2: @maci3jka please revert these formatting changes, they are not needed for this PR

> > to capi on [aws prerequisites documentation](https://cluster-api-aws.sigs.k8s.io/topics/using-clusterawsadm-to-fulfill-prerequisites#ssh-key-pair)

Reviewer: nit: CAPI, AWS

  *  Additional environment variables when testing cluster upgrades:
  ```bash
export CAPI_UPGRADE_VERSION=v1.28.0
export CAPI_UPGRADE_MD_NAME=${CLUSTER_NAME}-md-0
export CAPI_UPGRADE_MD_TYPE=machinedeployments.cluster.x-k8s.io
export CAPI_UPGRADE_CP_NAME=${CLUSTER_NAME}-control-plane
export CAPI_UPGRADE_CP_TYPE=microk8scontrolplanes.controlplane.cluster.x-k8s.io
# Change the control plane and worker machine count to desired values for in-place upgrades tests and create a new cluster manifest.
CONTROL_PLANE_MACHINE_COUNT=1
WORKER_MACHINE_COUNT=1
clusterctl generate cluster ${CLUSTER_NAME} --from "templates/cluster-template-aws.yaml" --kubernetes-version 1.27.0 > cluster-inplace.yaml
export CLUSTER_INPLACE_MANIFEST_FILE=$PWD/cluster-inplace.yaml

> ```

Reviewer 2: Revert, address in follow-up PR

  *  Additional environment variables when testing disable default CNI flag:
  ```bash
export DISABLE_DEFAULT_CNI=true

> export POST_RUN_COMMANDS='["helm install cilium cilium/cilium  --namespace kube-system --set cni.confPath=/var/snap/microk8s/current/args/cni-network --set cni.binPath=/var/snap/microk8s/current/opt/cni/bin --set daemon.runPath=/var/snap/microk8s/current/var/run/cilium --set operator.replicas=1 --set ipam.operator.clusterPoolIPv4PodCIDRList=\"10.1.0.0/16\" --set nodePort.enabled=true"]' # install Calico in place default CNI

Reviewer: why are you running this as a post run command?
I would just setup the cluster below and then do something like this:
```
	t.Run("DeployCluster", func(t *testing.T) { deployCluster(t, os.Getenv("CLUSTER_DISABLE_DEFAULT_CNI_MANIFEST_FILE")) })
	t.Run("ValidateCalico", func(t *testing.T) { validateCalico(t) })
	t.Run("DeployCilium", func(t *testing.T) { deployCilium(t) })
	t.Run("ValidateCilium", func(t *testing.T) { validateCilium(t) })
```

Reviewer 2: I don't think the test should depend on the user setting env. variables. This will be confusing in the future when you want to run the whole test suite and the test suite depends on a bunch of variables and you have to go hunt for them. Instead, create a manifests directory, and hard code the file there, and load it from the test.

clusterctl generate cluster ${CLUSTER_NAME} --from "templates/cluster-template-aws.yaml" --kubernetes-version 1.27.0 > cluster_disable_default_cni.yaml

> export CLUSTER_DISABLE_DEFAULT_CNI_MANIFEST_FILE=$PWD/cluster_disable_default_cni.yaml

Reviewer 2: See above, create manifests directory and load this file from there

The integration/e2e tests have the following prerequisites:

>   * an environment variable `CLUSTER_MANIFEST_FILE` pointing to the cluster manifest. Cluster manifests can be produced with the help of the templates found under `templates`. For example:

Author: yes those are not required they are hardcoded in cluster manifest dir

>   * make sure to have ssh key in aws `capi`in `us-east-1 region` if you do not have key refer
>     to CAPI on [AWS prerequisites documentation](https://cluster-api-aws.sigs.k8s.io/topics/using-clusterawsadm-to-fulfill-prerequisites#ssh-key-pair)

Reviewer 3: ```suggestion
  * ensure an AWS ssh key pair named `capi` exists according to the [CAPI Provider AWS documentation](https://cluster-api-aws.sigs.k8s.io/topics/using-clusterawsadm-to-fulfill-prerequisites#ssh-key-pair)
```

>   * to run tests locally you will need to have on path fallowing programs `clusterctl`, `kubectl`, `helm`

Reviewer 3: ```suggestion
  * local testing requires the following to be available in the PATH: `clusterctl`, `kubectl`, `helm`
```

```bash
kubectl apply -f bootstrap/bootstrap-components.yaml -f control-plane/control-plane-components.yaml
```
### Cluster definitions for e2e

> Cluster definition are stored in [`manifests`](./cluster-manifests) directory.

Reviewer 3: ```suggestion
Cluster definitions are stored in the [`manifests`](./cluster-manifests) directory.
```
